### PR TITLE
Fix to make web sockets close gracefully on server shutdown

### LINKED ISSUE #7538: [Issue]: WebSockets not closed gracefully
### Please describe your bug

When the server is stopped it sends a notification to all connected WebSocket clients, the message looks like this:

```json
{"MessageType":"ServerShuttingDown","MessageId":"d9fbbca3037344b689490d35ec2d4ed7","Data":""}
```

_(id is unique)_

Unfortunately it then just stops the server, ending the connection abruptly. It would be nice if the clients get properly disconnected.

I think this could be implemented in one of these places:

https://github.com/jellyfin/jellyfin/blob/c512e783b34fe67c523e1f71a930c23219669f14/Emby.Server.Implementations/ApplicationHost.cs#L1147-L1166

https://github.com/jellyfin/jellyfin/blob/c512e783b34fe67c523e1f71a930c23219669f14/Emby.Server.Implementations/Session/SessionManager.cs#L1359-L1364

### Jellyfin Version

10.8.0

### if other:

beta.1

### Environment

_No response_

### Jellyfin logs

_No response_

### FFmpeg logs

_No response_

### Please attach any browser or client logs here

🤷 WIP Kotlin SDK logs:

```
2022-03-27 18:35:02.429 13963-14157/? I/OkHttpWebsocketSession: Receiving (raw) message {"MessageType":"ServerShuttingDown","MessageId":"d9fbbca3037344b689490d35ec2d4ed7","Data":""}
2022-03-27 18:35:02.475 13963-14157/? W/OkHttpWebsocketSession: WebSocket has failed
    java.net.SocketException: Connection reset
        at java.net.SocketInputStream.read(SocketInputStream.java:215)
        at java.net.SocketInputStream.read(SocketInputStream.java:144)
        at okio.InputStreamSource.read(JvmOkio.kt:90)
        at okio.AsyncTimeout$source$1.read(AsyncTimeout.kt:129)
        at okio.RealBufferedSource.request(RealBufferedSource.kt:206)
        at okio.RealBufferedSource.require(RealBufferedSource.kt:199)
        at okio.RealBufferedSource.readByte(RealBufferedSource.kt:209)
        at okhttp3.internal.ws.WebSocketReader.readHeader(WebSocketReader.kt:119)
        at okhttp3.internal.ws.WebSocketReader.processNextFrame(WebSocketReader.kt:102)
        at okhttp3.internal.ws.RealWebSocket.loopReader(RealWebSocket.kt:293)
        at okhttp3.internal.ws.RealWebSocket$connect$1.onResponse(RealWebSocket.kt:195)
        at okhttp3.internal.connection.RealCall$AsyncCall.run(RealCall.kt:519)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
        at java.lang.Thread.run(Thread.java:920)
```

### Please attach any screenshots here

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by LewkyB**
**Changes**
- Added function `CloseAllWebSockets` to `SessionManager` that calls all the way down to the `WebSocketConnection` and uses `CloseOutputAsync` to gracefully close out the web socket when a system shutdown is called.
- Logs provided below to show the before and after log messages

**Issues**
Fixes #7538 

**Remarks**
- Not a huge fan of the nest for loops I did, maybe there's a better way?
- Should I change everything I did to `async`?
- This is my first time messing with a lot of this stuff. I'm very much looking forward to feedback!

**Kotlin SDK logs before:**
```
Starting subscription
[main] INFO org.jellyfin.sdk.api.client.KtorClient - POST http://localhost:8096/Users/AuthenticateByName
State DISCONNECTED
[DefaultDispatcher-worker-1] INFO org.jellyfin.sdk.api.sockets.SocketInstance - Reconnecting
State CONNECTING
[OkHttp http://localhost:8096/...] INFO org.jellyfin.sdk.api.sockets.OkHttpWebsocketSession - WebSocket has opened
[OkHttp http://localhost:8096/...] INFO org.jellyfin.sdk.api.sockets.OkHttpWebsocketSession - Receiving (raw) message {"MessageType":"ForceKeepAlive","MessageId":"00000000000000000000000000000000","Data":60}
[DefaultDispatcher-worker-2] INFO org.jellyfin.sdk.api.sockets.SocketInstance - Adding subscription for SessionsMessage
[DefaultDispatcher-worker-1] INFO org.jellyfin.sdk.api.sockets.helper.KeepAliveHelper - Using a KeepAlive message delay of 30 seconds
[DefaultDispatcher-worker-2] INFO org.jellyfin.sdk.api.sockets.OkHttpWebsocketSession - Sending (raw) message {"MessageType":"SessionsStart","Data":"0,1000"}
[DefaultDispatcher-worker-2] INFO org.jellyfin.sdk.api.sockets.OkHttpWebsocketSession - Sending (raw) message {"MessageType":"KeepAlive","Data":{}}
[DefaultDispatcher-worker-1] INFO org.jellyfin.sdk.api.sockets.SocketInstance - Adding subscription for ActivityLogEntryMessage
[OkHttp http://localhost:8096/...] INFO org.jellyfin.sdk.api.sockets.OkHttpWebsocketSession - Receiving (raw) message {"MessageType":"KeepAlive","MessageId":"eda0a22c5f8045e2bef3382736abbb3e"}
[DefaultDispatcher-worker-1] INFO org.jellyfin.sdk.api.sockets.SocketInstance - Adding subscription for ScheduledTasksInfoMessage
[DefaultDispatcher-worker-2] INFO org.jellyfin.sdk.api.sockets.OkHttpWebsocketSession - Sending (raw) message {"MessageType":"ActivityLogEntryStart","Data":"0,1000"}
State CONNECTED
[DefaultDispatcher-worker-2] INFO org.jellyfin.sdk.api.sockets.OkHttpWebsocketSession - Sending (raw) message {"MessageType":"ScheduledTasksInfoStart","Data":"0,1000"}
[DefaultDispatcher-worker-1] INFO org.jellyfin.sdk.api.sockets.OkHttpWebsocketSession - Sending (raw) message {"MessageType":"KeepAlive","Data":{}}
[OkHttp http://localhost:8096/...] INFO org.jellyfin.sdk.api.sockets.OkHttpWebsocketSession - Receiving (raw) message {"MessageType":"KeepAlive","MessageId":"791199526eb34384bd602bd91707b9a6"}
[OkHttp http://localhost:8096/...] INFO org.jellyfin.sdk.api.sockets.OkHttpWebsocketSession - Receiving (raw) message {"MessageType":"ServerShuttingDown","MessageId":"87116115715c4b219793656a5dd505dd","Data":""}
Received ServerShuttingDownMessage(messageId=87116115-715c-4b21-9793-656a5dd505dd)
[OkHttp http://localhost:8096/...] WARN org.jellyfin.sdk.api.sockets.OkHttpWebsocketSession - WebSocket has failed
java.io.EOFException
        at okio.RealBufferedSource.require(RealBufferedSource.kt:199)
        at okio.RealBufferedSource.readByte(RealBufferedSource.kt:209)
        at okhttp3.internal.ws.WebSocketReader.readHeader(WebSocketReader.kt:119)
        at okhttp3.internal.ws.WebSocketReader.processNextFrame(WebSocketReader.kt:102)
        at okhttp3.internal.ws.RealWebSocket.loopReader(RealWebSocket.kt:293)
        at okhttp3.internal.ws.RealWebSocket$connect$1.onResponse(RealWebSocket.kt:195)
        at okhttp3.internal.connection.RealCall$AsyncCall.run(RealCall.kt:519)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
        at java.base/java.lang.Thread.run(Thread.java:829)
State DISCONNECTED
```

**Kotlin SDK logs after:**
```
Starting subscription
[main] INFO org.jellyfin.sdk.api.client.KtorClient - POST http://localhost:8096/Users/AuthenticateByName
[DefaultDispatcher-worker-1] INFO org.jellyfin.sdk.api.sockets.SocketInstance - ReconnectingState DISCONNECTED

State CONNECTING
[OkHttp http://localhost:8096/...] INFO org.jellyfin.sdk.api.sockets.OkHttpWebsocketSession - WebSocket has opened
[OkHttp http://localhost:8096/...] INFO org.jellyfin.sdk.api.sockets.OkHttpWebsocketSession - Receiving (raw) message {"MessageType":"ForceKeepAlive","MessageId":"00000000000000000000000000000000","Data":60}
[DefaultDispatcher-worker-2] INFO org.jellyfin.sdk.api.sockets.SocketInstance - Adding subscription for SessionsMessage
[DefaultDispatcher-worker-2] INFO org.jellyfin.sdk.api.sockets.OkHttpWebsocketSession - Sending (raw) message {"MessageType":"SessionsStart","Data":"0,1000"}
[DefaultDispatcher-worker-2] INFO org.jellyfin.sdk.api.sockets.SocketInstance - Adding subscription for ActivityLogEntryMessage
[DefaultDispatcher-worker-1] INFO org.jellyfin.sdk.api.sockets.helper.KeepAliveHelper - Using a KeepAlive message delay of 30 seconds
[DefaultDispatcher-worker-2] INFO org.jellyfin.sdk.api.sockets.SocketInstance - Adding subscription for ScheduledTasksInfoMessage
[DefaultDispatcher-worker-1] INFO org.jellyfin.sdk.api.sockets.OkHttpWebsocketSession - Sending (raw) message {"MessageType":"ActivityLogEntryStart","Data":"0,1000"}
[DefaultDispatcher-worker-1] INFO org.jellyfin.sdk.api.sockets.OkHttpWebsocketSession - Sending (raw) message {"MessageType":"KeepAlive","Data":{}}
[OkHttp http://localhost:8096/...] INFO org.jellyfin.sdk.api.sockets.OkHttpWebsocketSession - Receiving (raw) message {"MessageType":"KeepAlive","MessageId":"3d76f92db07b4be9b631d710c33d9c00"}
State CONNECTED
[DefaultDispatcher-worker-1] INFO org.jellyfin.sdk.api.sockets.OkHttpWebsocketSession - Sending (raw) message {"MessageType":"ScheduledTasksInfoStart","Data":"0,1000"}
[OkHttp http://localhost:8096/...] INFO org.jellyfin.sdk.api.sockets.OkHttpWebsocketSession - WebSocket is closing, code=1000, reason=System Shutdown
State DISCONNECTED
[DefaultDispatcher-worker-1] INFO org.jellyfin.sdk.api.sockets.helper.ReconnectHelper - Reconnect scheduled in 5s (error=false)
[DefaultDispatcher-worker-1] INFO org.jellyfin.sdk.api.sockets.SocketInstance - Reconnecting
State CONNECTING
[OkHttp WebSocket http://localhost:8096/... writer] INFO org.jellyfin.sdk.api.sockets.OkHttpWebsocketSession - WebSocket has closed, code=1000, reason=System Shutdown
[OkHttp http://localhost:8096/...] WARN org.jellyfin.sdk.api.sockets.OkHttpWebsocketSession - WebSocket has failed
java.net.SocketTimeoutException: timeout
        at okio.SocketAsyncTimeout.newTimeoutException(JvmOkio.kt:143)
        at okio.AsyncTimeout.access$newTimeoutException(AsyncTimeout.kt:162)
        at okio.AsyncTimeout$source$1.read(AsyncTimeout.kt:335)
        at okio.RealBufferedSource.indexOf(RealBufferedSource.kt:427)
        at okio.RealBufferedSource.readUtf8LineStrict(RealBufferedSource.kt:320)
        at okhttp3.internal.http1.HeadersReader.readLine(HeadersReader.kt:29)
        at okhttp3.internal.http1.Http1ExchangeCodec.readResponseHeaders(Http1ExchangeCodec.kt:178)
        at okhttp3.internal.connection.Exchange.readResponseHeaders(Exchange.kt:106)
        at okhttp3.internal.http.CallServerInterceptor.intercept(CallServerInterceptor.kt:79)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
        at okhttp3.internal.connection.ConnectInterceptor.intercept(ConnectInterceptor.kt:34)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
        at okhttp3.internal.cache.CacheInterceptor.intercept(CacheInterceptor.kt:95)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
        at okhttp3.internal.http.BridgeInterceptor.intercept(BridgeInterceptor.kt:83)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
        at okhttp3.internal.http.RetryAndFollowUpInterceptor.intercept(RetryAndFollowUpInterceptor.kt:76)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
        at okhttp3.internal.connection.RealCall.getResponseWithInterceptorChain$okhttp(RealCall.kt:201)
        at okhttp3.internal.connection.RealCall$AsyncCall.run(RealCall.kt:517)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
        at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: java.net.SocketException: Socket closed
        at java.base/java.net.SocketInputStream.read(SocketInputStream.java:183)
        at java.base/java.net.SocketInputStream.read(SocketInputStream.java:140)
        at okio.InputStreamSource.read(JvmOkio.kt:90)
        at okio.AsyncTimeout$source$1.read(AsyncTimeout.kt:129)
        ... 20 more
```

### CODE REVIEW

File: Emby.Server.Implementations/Session/WebSocketController.cs [hunk 1]
```diff
@@ -88,6 +88,18 @@ private void OnConnectionClosed(object? sender, EventArgs e)
                 cancellationToken);
         }
 
+        /// <summary>
+        /// Gracefully closes all web sockets.
+        /// </summary>
+        /// <param name="cancellationToken">The cancellation token.</param>
```
Comment (crobibero) [inline Emby.Server.Implementations/Session/WebSocketController.cs:94]:
```suggestion
        /// <inheritdoc />
```

File: Emby.Server.Implementations/HttpServer/WebSocketConnection.cs [hunk 2]
```diff
@@ -231,6 +231,15 @@ private Task SendKeepAliveResponse()
                 CancellationToken.None);
         }
 
+        /// <summary>
+        /// Gracefully closes the socket.
+        /// </summary>
+        /// <param name="cancellationToken">The cancellation token.</param>
```
Comment (crobibero) [inline Emby.Server.Implementations/HttpServer/WebSocketConnection.cs:237]:
```suggestion
        /// <inheritdoc />
```

File: Emby.Dlna/PlayTo/PlayToController.cs [hunk 3]
```diff
@@ -889,6 +889,11 @@ public Task SendMessage<T>(SessionMessageType name, Guid messageId, T data, Canc
             return Task.CompletedTask;
         }
 
+        public void CloseAllWebSockets(CancellationToken cancellationToken)
```
Comment (crobibero) [inline Emby.Dlna/PlayTo/PlayToController.cs:892]:
```suggestion
		/// <inheritdoc />
        public void CloseAllWebSockets(CancellationToken cancellationToken)
```

File: Emby.Dlna/PlayTo/PlayToController.cs [hunk 4]
```diff
@@ -889,6 +889,11 @@ public Task SendMessage<T>(SessionMessageType name, Guid messageId, T data, Canc
...
+        {
+            throw new NotImplementedException();
```
Comment (crobibero) [inline Emby.Dlna/PlayTo/PlayToController.cs:894]: I think it's fine to just make this function do nothing rather then throwing an exception

File: Emby.Server.Implementations/HttpServer/WebSocketConnection.cs [hunk 5]
```diff
@@ -231,6 +231,15 @@ private Task SendKeepAliveResponse()
...
+        public void CloseSocket(CancellationToken cancellationToken)
+        {
+            _socket.CloseOutputAsync(WebSocketCloseStatus.NormalClosure, "System Shutdown", cancellationToken);
```
Comment (crobibero) [inline Emby.Server.Implementations/HttpServer/WebSocketConnection.cs:240]: This needs an `await`

**Review by crobibero — COMMENTED** [review]
Since this fixes a bug in 10.8 please rebase and target `release-10.8.z`

File: Emby.Server.Implementations/HttpServer/WebSocketConnection.cs (hunk 5 above)
Comment (LewkyB) [reply Emby.Server.Implementations/HttpServer/WebSocketConnection.cs]: Done, I also made everything in the call stack async too

File: Emby.Dlna/PlayTo/PlayToController.cs (hunk 4 above)
Comment (LewkyB) [reply Emby.Dlna/PlayTo/PlayToController.cs]: Done!

File: Emby.Server.Implementations/Session/WebSocketController.cs (hunk 1 above)
Comment (LewkyB) [reply Emby.Server.Implementations/Session/WebSocketController.cs]: Made the changes. Should've looked into what inheritdoc did! Thanks for taking a look at my stuff

**Comment by LewkyB** [comment]
Messed up the rebase! Sorry guys, opening a new PR with a clean history

**Pull request closed without merging**